[WIP]: Use `cibuildwheel` to build WASM/Pyodide wheels for `scikit-image`, push nightlies to Anaconda.org

agriyakhetarpal opened the pull request:
## Description

<!--
- Reference relevant issues or related pull requests with their URL / #<number>.
- Use `pre-commit` to check and format code.
-->

This PR is a follow-up to #7350 based on https://github.com/scikit-image/scikit-image/pull/7350#discussion_r1568713998. It modifies the wheel recipe workflow in `wheels_recipe.yml` to additionally build and test Pyodide wheels through `cibuildwheel` version 2.19.0, which added experimental support for building to a WebAssembly target. Since the workflow is triggered using the `workflow_call:` event in `nightly_wheel_build.yml`, this enables pushing the wheels to the Anaconda.org index, which is hosted at https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/

## Checklist

<!-- Before pull requests can be merged, they should provide: -->

- [x] A descriptive but concise pull request title
- [ ] [Docstrings for all functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [ ] [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- [ ] A gallery example in `./doc/examples` for new features
- [ ] [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
...
```

agriyakhetarpal commented:
This isn't currently working because the Node.js script runner isn't able to find `pyodide` where it gets installed: https://github.com/agriyakhetarpal/scikit-image/actions/runs/9471953221/job/26096398007

I think I have two potential fixes; one is to do `npm install pyodide --global` (this didn't work), and the other is to install to the `{project}` directory, i.e., the root of the repository, as a prefix because that is where I have instructed `cibuildwheel` to run the JavaScript file from. I'll test this in a moment.

Summoning @hoodmane – would you consider this a bug or rather an intricacy of how `npm` works (I am not proficient with JavaScript enough)? In any case, I would like to improve the Pyodide docs around all of this – not just for highlighting how to configure/use `cibuildwheel` with it, but also for out-of-tree builds in general and recommendations around it, in the coming time. I still have some other documentation work from our previous discussions in my backlog :)

hoodmane commented:
I just noticed your ping here @agriyakhetarpal I'll try to get back to you about it soon.

agriyakhetarpal commented:
This can now be revisited once Pyodide 0.27 stable is released at the end of the month, and once `cibuildwheel` is updated to use it.

agriyakhetarpal commented:
> This can now be revisited once Pyodide 0.27 stable is released at the end of the month, and once `cibuildwheel` is updated to use it.

`cibuildwheel` has been updated to use Pyodide 0.27 through its v2.23.0 release today. I've updated the PR description and dropped all unneeded changes post #7525. The wheel build in `emscripten.yaml` does not use `cibuildwheel` however, as `cibuildwheel` can take some time to update its internal Pyodide version, as the Pyodide version is coupled with its version. This is ready for review again. Thanks!

agriyakhetarpal commented on .github/workflows/wheels-recipe.yaml:
```diff
@@ -17,6 +17,7 @@ on:
         required: false
         type: string
         default: "pytest --pyargs skimage"
+  workflow_dispatch: # temporarily added for debugging
```
Adding a placeholder comment to remove it, once we're ready.
```suggestion
```

agriyakhetarpal commented:
The workflow run passes with `cibuildwheel`; here are the logs from my fork: https://github.com/agriyakhetarpal/scikit-image/actions/runs/13605820521/job/38037164891?pr=2

stefanv commented on .github/workflows/wheels-recipe.yaml:
```diff
@@ -253,3 +253,26 @@ jobs:
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
```
Fetch depth required?

agriyakhetarpal replied:
I'm not sure if it is, but I saw other jobs in this file using it, so I decided to keep it.

stefanv replied:
Check if you need it: https://github.com/actions/checkout

agriyakhetarpal replied:
AFAICS, we aren't using any tool that requires anything but a shallow clone; will drop it.

agriyakhetarpal replied:
Addressed via 1ba76682cdd233c56fa4a5a8c2884ad2ce0fe712

stefanv reviewed (approved):
Looks good to me!

Maybe add a note to emscripten to explain the difference in build methodology?

agriyakhetarpal commented:
> Maybe add a note to emscripten to explain the difference in build methodology?

Yes, sure; where should I add this? Do you mean adding one above the workflow?

stefanv commented:
> > Maybe add a note to emscripten to explain the difference in build methodology?
> 
> Yes, sure; where should I add this? Do you mean adding one above the workflow?

Sure, anywhere to document that there is a discrepancy in the way we build emscripten, and why. Probably should be mentioned in both builds!

agriyakhetarpal commented:
Got it! I won't say there's a difference in how we build for Emscripten. The only changes from a regular `pip install build && python -m build` are:
- we use `pyodide build` which monkeypatches `pypa/build` and uses `pywasmcross.py` script to cross-compile from Linux to wasm32-unknown-emscripten, instead of compiling natively
- cibuildwheel uses `pyodide build` as well and sets up Emscripten, Node.js, etc. automatically
- the testing differs by a bit as `pytest`'s cacheprovider plugin is somewhat broken – which I noted using a comment.

This translates to just one extra option to pass to `cibuildwheel`, which can only be configured using an environment variable at the moment and is done using `CIBW_PLATFORM: pyodide`.

Or, were you referring to how this job and the workflow in `emscripten.yaml` differ from each other?

stefanv commented:
Ah, OK, I just wasn't sure of the reasoning of why there was any difference at all, and wanted to make sure others who look into this in the future can follow. However you choose to document that is fine.

agriyakhetarpal commented:
Okay, I added a note about this in 49ac338172e1ec509ce262e179915ab3e290fdeb. The git blame should also point to this PR so developers can examine why the difference exists. Thanks for the review!

agriyakhetarpal commented:
I addressed the merge conflicts here. Thanks!

I notice that `zizmor` did not warn me about missing SHAs for the `cibuildwheel` action, as it's not running with the `--pedantic` argument. I assume it was removed at some later point, but I've still added the hash.

agriyakhetarpal commented:
I'm having a bit of trouble running the tests on my fork to see if this is working: https://github.com/agriyakhetarpal/scikit-image/actions/runs/14802982173

I'll debug this, and we should be good to proceed once we have a passing wheel build.

stefanv commented:
Thanks, @agriyakhetarpal!

We're having a chat with the Anaconda team soon and will mention the CORS header issue so that, hopefully, we can start using the dev wheels from the nightly repo directly.

agriyakhetarpal commented:
Thanks for this, @stefanv ! Actually, the CORS headers issue from the Anaconda.org side has been resolved earlier this year (see pyodide/pyodide#4898#issuecomment-2637731077), so these wheels are now usable in the interactive docs for the "dev" versions. Here's an example from `scikit-learn`: scikit-learn/scikit-learn#31085

stefanv commented:
@matthewfeickert ☝️
